File: GLLara/GLLModel.m
//
//  GLLModel.m
//  GLLara
//
//  Created by Torsten Kammer on 31.08.12.
//  Copyright (c) 2012 Torsten Kammer. All rights reserved.
//

#import "GLLModel.h"

#import <UniformTypeIdentifiers/UniformTypeIdentifiers.h>

#import "NSArray+Map.h"
#import "GLLASCIIScanner.h"
#import "GLLModelObj.h"
#import "TRInDataStream.h"
#import "TROutDataStream.h"

#import "GLLara-Swift.h"

NSString *GLLModelLoadingErrorDomain = @"GLL Model loading error domain";

static NSCache *cachedModels;

@implementation GLLModel

+ (void)initialize
{
    cachedModels = [[NSCache alloc] init];
}

+ (id)cachedModelFromFile:(NSURL *)file parent:(GLLModel *)parent error:(NSError *__autoreleasing*)error;
{
    id key = file.absoluteString;
    if (parent != nil)
        key = [file.absoluteString stringByAppendingFormat:@"\n%@", parent.baseURL.absoluteString];
    
    id result = [cachedModels objectForKey:key];
    if (!result)
    {
        if ([file.path hasSuffix:@".mesh"] || [file.path hasSuffix:@".xps"])
        {
            result = [[GLLModelXNALara alloc] initWithBinaryFromFile:file parent:parent error:error];
            if (!result) return nil;
        }
        else if ([file.path hasSuffix:@".mesh.ascii"])
        {
            result = [[GLLModelXNALara alloc] initWithASCIIFromFile:file parent:parent error:error];
            if (!result) return nil;
        }
        else if ([file.path hasSuffix:@".obj"])
        {
            result = [[GLLModelObj alloc] initWithContentsOfURL:file error:error];
            if (!result) return nil;
        }
        else if ([file.path hasSuffix:@".gltf"])
        {
            result = [[GLLModelGltf alloc] initWithUrl:file isBinary:NO error:error];
            if (!result) return nil;
        }
        else if ([file.path hasSuffix:@".glb"])
        {
            result = [[GLLModelGltf alloc] initWithUrl:file isBinary:YES error:error];
            if (!result) return nil;
        }
        else
        {
            if (error)
            {
                // Find display name for this extension
                UTType *fileType;
                if (![file getResourceValue:&fileType forKey:NSURLContentTypeKey error:error]) {
                    return nil;
                }
                NSString *fileTypeDescription = fileType.localizedDescription;
                
                *error = [NSError errorWithDomain:GLLModelLoadingErrorDomain code:GLLModelLoadingError_FileTypeNotSupported userInfo:@{
                                                                                                                                       NSLocalizedDescriptionKey : [NSString stringWithFormat:NSLocalizedString(@"Files of type %@ are not supported.", @"Tried to load unsupported format"), fileTypeDescription],
                                                                                                                                       NSLocalizedRecoverySuggestionErrorKey : NSLocalizedString(@"Only .mesh, .mesh.ascii and .obj files can be loaded.", @"Tried to load unsupported format")}];
            }
            return nil;
        }
        
        [cachedModels setObject:result forKey:key];
    }
    return result;
}

#pragma mark - Accessors

- (BOOL)hasBones
{
    return self.bones.count > 0;
}

- (NSArray<GLLModelBone *> *)rootBones
{
    return [self.bones filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"hasParent == NO"]];
}

- (NSArray<GLLCameraTargetDescription *> *)cameraTargetNames
{
    return self.parameters.cameraTargets;
}

- (GLLModelBone *)boneForName:(NSString *)name;
{
    return [self.bones firstObjectMatching:^BOOL(GLLModelBone *bone){
        return [bone.name isEqual:name];
    }];
}

@end

File: GLLara/GLLModel.swift
//
//  GLLModel.m
//  GLLara
//
//  Created by Torsten Kammer on 31.08.12.
//  Copyright (c) 2012 Torsten Kammer. All rights reserved.
//

import Foundation
import UniformTypeIdentifiers

let GLLModelLoadingErrorDomain = "GLL Model loading error domain";

enum GLLModelLoadingErrorCode: Int {
    case prematureEndOfFile
    case indexOutOfRange
    case circularReference
    case fileTypeNotSupported
    case parametersNotFound
}

/**
 * # A renderable object.
 *
 * A GLLModel corresponds to one mesh file (which actually contains many meshes; this is a bit confusing) and describes its graphics contexts. It contains some default transformations, but does not store poses and the like.
 */
@objc class GLLModel: NSObject {
    @objc var bones: [GLLModelBone] = []
    @objc var meshes: [GLLModelMesh] = []
    
    static let cachedModels = NSCache<NSString, GLLModel>()
    
    /**
     * # Returns a model with a given URL, returning a cached instance if one exists.
     *
     * Since a model is immutable here, it can be shared as much as necessary. This method uses an internal cache to share objects. Note that a model can be evicted from this cache again, if nobody is using it.
     */
    @objc static func cachedModel(from file: URL, parent: GLLModel? = nil) throws -> GLLModel {
        var key = file.absoluteString
        if let parent {
            key += "\n\(parent.baseURL.absoluteString)"
        }
        if let result = cachedModels.object(forKey: key as NSString) {
            return result
        }
        
        if file.pathExtension == "mesh" || file.pathExtension == "xps" {
            let model = try GLLModelXNALara(binaryFromFile: file, parent: parent)
            cachedModels.setObject(model, forKey: key as NSString)
            return model
        } else if file.lastPathComponent.hasSuffix(".mesh.ascii") {
            let model = try GLLModelXNALara(ASCIIFromFile: file, parent: parent)
            cachedModels.setObject(model, forKey: key as NSString)
            return model
        } else if file.pathExtension == "obj" {
            let model = try GLLModelObj(contentsOf: file)
            cachedModels.setObject(model, forKey: key as NSString)
            return model
        } else if file.pathExtension == ".gltf" {
            let model = try GLLModelGltf(url: file, isBinary: false)
            cachedModels.setObject(model, forKey: key as NSString)
            return model
        } else if file.pathExtension == ".glb" {
            let model = try GLLModelGltf(url: file, isBinary: true)
            cachedModels.setObject(model, forKey: key as NSString)
            return model
        } else {
            // Find display name for this extension
            let contentType = try file.resourceValues(forKeys: [.contentTypeKey]).contentType
            let fileTypeDescription = contentType?.localizedDescription ?? file.pathExtension
            
            throw NSError(domain: GLLModelLoadingErrorDomain, code: GLLModelLoadingErrorCode.fileTypeNotSupported.rawValue, userInfo: [
                NSLocalizedDescriptionKey : String(format:NSLocalizedString("Files of type %@ are not supported.", comment: "Tried to load unsupported format"), fileTypeDescription),
                NSLocalizedRecoverySuggestionErrorKey : NSLocalizedString("Only .mesh, .mesh.ascii and .obj files can be loaded.", comment: "Tried to load unsupported format")
            ])

        }
    }

    @objc var baseURL: URL! = nil
    @objc var parameters: GLLModelParams! = nil
    
    @objc var hasBones: Bool {
        return bones.count > 0
    }
    
    @objc var rootBones: [GLLModelBone] {
        return bones.filter({ bone in bone.parent == nil } as (GLLModelBone)->Bool)
    }
    
    @objc var cameraTargetNames: [GLLCameraTargetDescription] {
        return parameters.cameraTargets
    }
    
    @objc func bone(name: String) -> GLLModelBone? {
        return bones.first { $0.name == name }
    }
}
